kyo-prelude/shared/src/main/scala/kyo/Batch.scala

```diff
@@ -108,62 +108,63 @@ object Batch:
+        def expand(state: Seq[ToExpand | Expanded | A]): Chunk[Expanded | A] < (S & S2) =
             Kyo.foreach(state) {
-                case Cont(Op.Eval(seq), cont) =>
+                case ToExpand(seq: Seq[Any], cont) =>
                     Kyo.foreach(seq)(v => runCont(cont(v))).map(expand)
-                case Cont(Op.Call(v, source), cont) =>
-                    Seq(Expanded(v, source.asInstanceOf, cont))
-                case a: A @unchecked =>
-                    Seq(a)
-            }.map(_.flatten)
+                case expanded: Expanded @unchecked => Chunk(expanded)
+                case done: A @unchecked            => Chunk(done)
+            }.map(_.flattenChunk)
 
-        def loop(state: Seq[Cont | A]): Seq[A] < (S & S2) =
+        def loop(state: Seq[ToExpand | Expanded | A]): Seq[A] < (S & S2) =
             expand(state).map { expanded =>
-                val (completed, pending) =
-                    expanded.zipWithIndex.partitionMap {
-                        case (e: Expanded @unchecked, idx) => Right((e, idx))
-                        case (a: A @unchecked, idx)        => Left((a, idx))
-                    }
-
-                if pending.isEmpty then
-                    completed.sortBy(_._2).map(_._1)
+                if !expanded.exists((_: @unchecked).isInstanceOf[Expanded]) then
+                    expanded.asInstanceOf[Chunk[A]]
                 else
-                    val grouped = pending.groupBy(_._1.source)
-
-                    Kyo.foreach(grouped.toSeq) { case (source, items) =>
-                        val (expandedItems, indices) = items.unzip
-                        val values                   = expandedItems.map(_.v)
-                        val uniqueValues             = values.distinct
-
-                        source(uniqueValues).map { map =>
-                            Kyo.foreach(expandedItems.zip(indices)) { case (e, idx) =>
-                                runCont(map(e.v).map(e.cont)).map((_, idx))
-                            }
+                    val pending: Map[SourceAny | Unit, Seq[Expanded | A]] =
+                        expanded.groupBy {
+                            case Expanded(_, source, _) => source
+                            case _                      => ()
                         }
+                    Kyo.foreach(pending.toSeq) {
+                        case (_: Unit, items) => Chunk.from(items)
+                        case (source: SourceAny, items: Seq[Expanded] @unchecked) =>
+                            source(items.map(_.value).distinct).map { map =>
+                                Kyo.foreach(items) { e =>
+                                    runCont(map(e.value).map(e.cont))
+                                }
+                            }
                     }.map { results =>
-                        loop(completed.map(_._1) ++ results.flatten.map(_._1))
+                        loop(results.flattenChunk)
```

Comment:
The impl had some logic to ensure ordering is preserved but this line was not considering it. I've decided to remove the ordering mechanism because the execution itself already produces results in the expected order.